### PR TITLE
add Dockerfile and docker-compose to containerize application

### LINKED ISSUE #696: Dockerize the miru application.
### Is your proposal related to a problem?

Setting up the Miru app is kind of a crazy process, coz it has a lot of package dependencies and a couple of developers faced issues with setup the app in their environment (Specifically non-Unix, eg: Windows). So It's better to do a dockerize the application, then it is easy to set up in any environment.

### Describe the solution you'd like

Create a docker-composer file and add all the dependencies to it.

### PR DESCRIPTION
**Pull request by rajat-tomar**
Fixes #696 

## Summary
Dockerfile file and docker-compose has been added to run application in local development

* To run the app in local:
  * Stop local services like postgresql, elasticsearch and redis if they are running
  * clone the repo
  * go to the app directory
  * run docker compose up
    * this will build the image and run the services defined in docker-compose. Webpacker will compile the assets and server will start with postgres, redis, sidekiq and elasticsearch
    * once the app is running, to setup database you can either run 
      ```docker-compose exec app bundle exec rails db:reset``` or
      ```docker-compose run --rm app bundle exec rails db:reset```
    The latter can be used even if the app is not running. This will start the service in container and will setup the database removing previous data if there is any.
Note that ```db:reset``` and ```db:setup``` both will work here.
  * To run rails console:
    * When app is running: ```docker-compose exec app bundle exec rails console``` or
When app is not running ```docker-compose run --rm app bundle exec rails console```
  * Similarly if you add new migrations or want to rollback previous one you can always follow the above mentioned procedure like:
    * ```docker-compose exec app bundle exec rails db:migrate``` when app service is running
    * and ```docker-compose run --rm app bundle exec rails db:migrate``` when service is not running

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

### Checklist:

- [x] I have manually tested all workflows
- [x] I have performed a self-review of my own code

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,49 @@
+#syntax=docker/dockerfile:1
```
Comment (akhilgkrishnan) [inline Dockerfile:1]: @rajat-tomar can you rename the `Dockerfile` to `Dockerfile.local`.

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      - REDIS_URL=redis://redis:6379/1
+      - ELASTICSEARCH_URL=http://elasticsearch:9200
+    depends_on:
+      - database
+
+  sidekiq:
+    build:
+      context: .
+      dockerfile: Dockerfile
+    command: bundle exec sidekiq -e development
```
Comment (akhilgkrishnan) [inline docker-compose.yml:54]:
```suggestion
    command: bundle exec sidekiq -C config/sidekiq.yml
```

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,79 @@
+version: "3.9"
+services:
+  database:
+    image: postgres:13.8
```
Comment (akhilgkrishnan) [inline docker-compose.yml:4]:
```suggestion
    image: postgres:13.8
    restart: always
```

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    ports:
+      - "5432:5432"
+    volumes:
+      - db_data:/var/lib/postgresql/data
```
Comment (akhilgkrishnan) [inline docker-compose.yml:8]:
```suggestion
      - postgres:/var/lib/postgresql/data
```

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    environment:
+      discovery.type: single-node
+      network.host: 0.0.0.0
+      xpack.security.enabled: false
+    ports:
+      - "9200:9200"
+
+volumes:
+  gem_cache:
+  db_data:
```
Comment (akhilgkrishnan) [inline docker-compose.yml:78]:
```suggestion
  postgres:
```

File: docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    image: postgres:13.8
+    ports:
+      - "5432:5432"
+    volumes:
+      - db_data:/var/lib/postgresql/data
+    environment:
+      POSTGRES_PASSWORD: postgres
+
+  redis:
+    image: redis:6.2.7
```
Comment (akhilgkrishnan) [inline docker-compose.yml:13]:
```suggestion
    image: redis:6.2.7
    command: redis-server
     restart: always
     ports:
        - '6379:6379'
     
       
```

File: docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+      discovery.type: single-node
+      network.host: 0.0.0.0
+      xpack.security.enabled: false
+    ports:
+      - "9200:9200"
+
+volumes:
+  gem_cache:
+  db_data:
+  node_modules:
```
Comment (akhilgkrishnan) [inline docker-compose.yml:79]: can you also include volume for Redis as well

**Review by akhilgkrishnan — COMMENTED** [review]
@rajat-tomar Added a couple of suggestions, for the rest of LGTM. Once the PR is merged can you also write documentation for the docker setup?

**Comment by akhilgkrishnan** [comment]
@apoorv1316 Can you try to setup this on your local?

File: docker-compose.yml (hunk 7 above)
Comment (rajat-tomar) [reply docker-compose.yml:87]: Hey @akhilgkrishnan, just wanted to ask you that why do we need volume for redis? Do we really need to persist the sidekiq job cache?

**Comment by rajat-tomar** [comment]
> @rajat-tomar Added a couple of suggestions, for the rest of LGTM. Once the PR is merged can you also write documentation for the docker setup?

Should I update the documentation in this PR itself or raise a new PR. If latter can you please open a new issue for the same.

**Comment by akhilgkrishnan** [comment]
> > @rajat-tomar Added a couple of suggestions, for the rest of LGTM. Once the PR is merged can you also write documentation for the docker setup?
> 
> Should I update the documentation in this PR itself or raise a new PR. If latter can you please open a new issue for the same.

@rajat-tomar We are working on a docusaurus app for Miru documentation. So there is a separate Github repo for that https://github.com/saeloun/miru-docs . Once we merged this PR, I can create a new issue on that repo. So you can add the documentation there.

**Comment by rajat-tomar** [comment]
> @apoorv1316 Can you try to setup this on your local?

Yeah, I've tested it and It is working fine in local

File: docker-compose.yml (hunk 7 above)
Comment (rajat-tomar) [reply docker-compose.yml:87]: @akhilgkrishnan Did you come to any conclusion yet? Do we add volume for redis?

Comment (akhilgkrishnan) [reply docker-compose.yml:87]: Yes, we need it. Coz sometimes we need to track the failed job. So it is better to keep it as persistent storage.

**Comment by akhilgkrishnan** [comment]
@rajat-tomar For the next time can you follow the git-flow pattern for creating the branching. eg: `feature/<Issue No>-<issue-name>`

**Comment by rajat-tomar** [comment]
> @rajat-tomar For the next time can you follow the git-flow pattern for creating the branching. eg: `feature/<Issue No>-<issue-name>`

Yeah sure, I'll keep that in mind!

**Review by akhilgkrishnan — APPROVED** [review]
LGTM

**Comment by akhilgkrishnan** [comment]
@prasanthchaduvula can you run this on your local

**Comment by akhilgkrishnan** [comment]
@rajat-tomar Thanks for working on this. I've created a new issue https://github.com/saeloun/miru-docs/issues/10 for documentation. If possible can you work on that.

**Pull request merged**